<a href="https://colab.research.google.com/github/danemiles/projects/blob/main/Fiddler_Sep_29%2C_2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **This Week’s Fiddler**

From Bowen Kerins comes a puzzle about pinball:

You’re playing a game of pinball that includes four lanes, each of which is initially unlit. Every time you flip the pinball, it passes through exactly one of the four lanes and toggles that lane’s state. So if that lane is unlit, it becomes lit after the ball passes through. But if the lane is lit, it becomes unlit after the ball passes through.

On average, how many times will you have to flip the pinball until all four lanes are lit?

In [119]:
# BACKGROUND ASSUMPTIONS:
# 1. You cannot "aim" for a lane. I.e. all lanes have equal probability of
#    being passed through.
# 2. Due to above, a lane is always at risk of being unlit.
# 3. You can only flip one lane per flip. I.e. you cannot flip,
#    the ball passes through a lane, the ball falls back passing through a
#    lane again, and finally returns to the flipper(s).

# ----------------------------------------------------
# import needed packages
import numpy as np
import pandas as pd
import random

# EDITABLE VARIABLES:
lane_count_values = np.arange(1,10) # range of individual lanes to test for
N = 100000  # Number of simulations

# ----------------------------------------------------

def flip(count, lanes):
  light_count = count
  lanes = lanes
  change = random.randint(0,light_count-1)
  if lanes[change] == 0:
    lanes[change] = 1
  else:
    lanes[change] == 0
  return lanes



In [120]:
# cool, this works like intended.
# let's add a second recursive function to save the result, run until solved, and write results to Pandas

def run_until_all_lit(count):
    lights = [0] * count
    flips = 0

    while sum(lights) < count:
        lights = flip(count, lights)
        flips += 1

    return flips

def simulate_game_N_times(N, count):
    total_flips = 0

    for _ in range(N):
        flips_needed = run_until_all_lit(count)
        total_flips += flips_needed

    average_flips = total_flips / N
    return average_flips

dfs = []

for count in lane_count_values:
    average_flips = simulate_game_N_times(N, count)
    df = pd.DataFrame({'Lane Count': [count], 'Average Flips': [average_flips]})
    dfs.append(df)

results_df = pd.concat(dfs, ignore_index=True)

results_df['Ratio'] = results_df['Average Flips'].shift(-1) / results_df['Average Flips']

# Print the DataFrame
print(results_df)

   Lane Count  Average Flips     Ratio
0           1        1.00000  3.007890
1           2        3.00789  1.830243
2           3        5.50517  1.515790
3           4        8.34468  1.370589
4           5       11.43713  1.287463
5           6       14.72488  1.234595
6           7       18.17927  1.195793
7           8       21.73864  1.171648
8           9       25.47004       NaN


### *UPSHOT: It will take approximately 8.34 flips to light all 4-lanes when the initial state is all unlit.*

---
# **Extra Credit**

Instead of four lanes, now suppose your pinball game has N lanes. And let’s say that E(N) represents the average number of pinball flips it takes until all N lanes are lit up.

Now, each time you increase the number of lanes by one, you find that it takes you approximately twice as long to light up all the lanes. In other words, E(N+1) seems to be about double E(N).

But upon closer examination, you find that it’s not quite double. Moreover, there’s a particular value of N where the ratio E(N+1)/E(N) is at a minimum. What is this value of N?

### *This is a bit trickier. At high values of N, recursion becomes impractical due to compute time. However, there does appear to be a tapering off, perhaps minimizing near E(N+1)/E(N) =1. Without a precise mathematical function / formula, I do not have a method to solve.*